### PR TITLE
New chart features

### PR DESCRIPTION
**Pull request by NikkiMeganMoore**

### CODE REVIEW

File: example/src/App.js [hunk 1]
```diff
@@ -521,9 +527,9 @@ export default class App extends Component {
   }
 
   onError = (error) => {
-    if (error && error.message && this.state.isAuthenticated) {
-      message.error(`${error.message}`)
-    }
+    // if (error && error.message && this.state.isAuthenticated) {
+    //   message.error(`${error.message}`)
+    // }
```
Comment (Ben-CHATA) [inline example/src/App.js:532]: Are we hanging onto this to be changed later?

File: example/src/App.js [hunk 2]
```diff
@@ -1193,23 +1207,42 @@ export default class App extends Component {
...
               // height: 'auto',
               // minHeight: '100px',
-              height: 'calc(100vh - 120px)',
+              height: 'calc(100vh - 140px)',
               overflow: 'hidden',
-              padding: '20px',
-              paddingTop: '0',
+              // padding: '20px',
+              // paddingTop: '0',
+              // paddingBottom: '0',
```
Comment (Ben-CHATA) [inline example/src/App.js:1229]: lots of commented styles here

File: src/components/Charts/Axis/Axis.scss [hunk 3]
```diff
@@ -30,6 +36,193 @@ g.legendOrdinal tspan {
...
+  color: var(--react-autoql-text-color-primary);
+  padding: 5px;
+  z-index: 9999;
+
+  .string-select-list-item {
+    font-size: 12px;
+    line-height: 28px;
+
+    &.active {
+      // background: rgba(0, 0, 0, 0.04);
```
Comment (Ben-CHATA) [inline src/components/Charts/Axis/Axis.scss:219]: cleanup

File: src/components/Charts/Axis/Axis.test.js [hunk 4]
```diff
@@ -44,10 +51,85 @@ describe('renders correctly', () => {
...
+        const wrapper = setup({
+          ...pivotSampleProps,
+          col: pivotSampleProps.columns[pivotSampleProps.stringColumnIndex],
+          scale: pivotSampleProps.stringScale({
+            title: pivotSampleProps.columns[pivotSampleProps.stringColumnIndex].display_name,
+          }),
+          orient: 'Left',
+        })
+
+        // const xLabel = findByTestAttr(wrapper, 'x-axis-label')
```
Comment (Ben-CHATA) [inline src/components/Charts/Axis/Axis.test.js:77]: cleanup

File: src/components/Charts/Axis/RowNumberSelector.js [hunk 5]
```diff
@@ -158,7 +161,7 @@ export default class RowNumberSelector extends React.Component {
         padding={10}
       >
         <rect
-          {...this.props.childProps}
+          // {...this.props.childProps}
```
Comment (Ben-CHATA) [inline src/components/Charts/Axis/RowNumberSelector.js:164]: cleanup

File: src/components/Charts/ChataChart/ChataChart.scss [hunk 6]
```diff
@@ -18,24 +26,124 @@
...
+        text {
+          color: var(--react-autoql-background-disabled);
+          opacity: 0.4 !important;
+        }
+        line.swatch,
+        path.swatch {
+          // fill-opacity: 0.1 !important;
+          // stroke-opacity: 0.1 !important;
+          // fill: var(--react-autoql-text-color-primary) !important;
+          // stroke: var(--react-autoql-text-color-primary) !important;
```
Comment (Ben-CHATA) [inline src/components/Charts/ChataChart/ChataChart.scss:45]: cleanup?

File: src/components/Charts/ChataChart/ChataChart.scss [hunk 7]
```diff
@@ -18,24 +26,124 @@
...
+      // &.visible {
+      //   text {
+      //     opacity: 1;
+      //   }
+      //   line.swatch,
+      //   path.swatch {
+      //     fill-opacity: 1 !important;
+      //     stroke-opacity: 1 !important;
+      //   }
+      // }
```
Comment (Ben-CHATA) [inline src/components/Charts/ChataChart/ChataChart.scss:60]: cleanup

File: src/components/Charts/Legend/Legend.js [hunk 8]
```diff
@@ -1,266 +1,385 @@
...
+                childProps={{
+                  ref: (r) => (this.legendBorder = r),
+                  x: (this.titleBBox?.x ?? 0),
+                  y: (this.titleBBox?.y ?? 0),
+                  width: (this.titleBBox?.width ?? 0) + this.BUTTON_PADDING * 2,
+                  height: (this.titleBBox?.height ?? 0) + this.BUTTON_PADDING * 2,
+                  transform: this.props.translate,
+                }}
+              />
+            )} */}
```
Comment (Ben-CHATA) [inline src/components/Charts/Legend/Legend.js:355]: cleanup?

File: src/components/Charts/chartTestData.js [hunk 9]
```diff
@@ -141,40 +144,124 @@ const datePivotData = [
...
+        columnIndex: listProps.stringColumnIndex,
+        domain: listData.map((row) => row[listProps.stringColumnIndex]),
+        ...params,
+      })
+      // const scale = scaleBand()
+      // scale
+      //   .domain(listData.map((d) => d[0]))
+      //   .range([0, 500])
+      //   .paddingInner(1)
+      //   .paddingOuter(0)
```
Comment (Ben-CHATA) [inline src/components/Charts/chartTestData.js:198]: cleanup?

File: src/components/Charts/chartTestData.js [hunk 10]
```diff
@@ -141,40 +144,124 @@ const datePivotData = [
...
+    },
+    numberScale: (params = {}) => {
+      const scale = getLinearScales({
+        props: listProps,
+        columnIndices1: listProps.numberColumnIndices,
+        isScaled: false,
+        ...params,
+      })?.scale
+      // const scale = scaleLinear()
+      // scale.domain([5000000, 20000000]).range([0, 500]).nice()
```
Comment (Ben-CHATA) [inline src/components/Charts/chartTestData.js:216]: cleanup

File: src/components/Charts/chartTestData.js [hunk 11]
```diff
@@ -141,40 +144,124 @@ const datePivotData = [
...
-      .paddingOuter(0),
-    numberScale: scaleLinear().domain([5000000, 20000000]).range([0, 500]).nice(),
+    ...pivotProps,
+    stringScale: (params = {}) => {
+      // const scale = scaleBand()
+      // scale
+      //   .domain(pivotData.map((d) => d[0]))
+      //   .range([0, 500])
+      //   .paddingInner(1)
+      //   .paddingOuter(0)
```
Comment (Ben-CHATA) [inline src/components/Charts/chartTestData.js:235]: cleanup

File: src/components/Charts/chartTestData.js [hunk 12]
```diff
@@ -141,40 +144,124 @@ const datePivotData = [
...
+      if (scale && params && Object.keys(params)?.length) {
+        Object.keys(params).forEach((key) => {
+          scale[key] = params[key]
+        })
+      }
+      return scale
+    },
+    numberScale: (params = {}) => {
+      // const scale = scaleLinear()
+      // scale.domain([5000000, 20000000]).range([0, 500]).nice()
```
Comment (Ben-CHATA) [inline src/components/Charts/chartTestData.js:251]: cleanup

File: src/components/Charts/chartTestData.js [hunk 13]
```diff
@@ -225,26 +312,48 @@ export default {
...
-      .paddingOuter(0),
-    numberScale: scaleLinear().domain([0, 1000]).range([0, 500]).nice(),
+    ...datePivotProps,
+    stringScale: (params = {}) => {
+      // const scale = scaleBand()
+      // scale
+      //   .domain(datePivotData.map((d) => d[0]))
+      //   .range([0, 500])
+      //   .paddingInner(1)
+      //   .paddingOuter(0)
```
Comment (Ben-CHATA) [inline src/components/Charts/chartTestData.js:326]: cleanup

File: src/components/Charts/chartTestData.js [hunk 14]
```diff
@@ -225,26 +312,48 @@ export default {
...
+      if (scale && params && Object.keys(params)?.length) {
+        Object.keys(params).forEach((key) => {
+          scale[key] = params[key]
+        })
+      }
+      return scale
+    },
+    numberScale: (params = {}) => {
+      // const scale = scaleLinear()
+      // scale.domain([0, 1000]).range([0, 500]).nice()
```
Comment (Ben-CHATA) [inline src/components/Charts/chartTestData.js:342]: cleanup

File: src/components/ChatMessage/ChatMessage.js [hunk 15]
```diff
@@ -285,7 +277,7 @@ export default class ChatMessage extends React.Component {
           rebuildTooltips={this.props.rebuildTooltips}
           source={this.props.source}
           onRowChange={this.scrollIntoView}
-          onDisplayTypeChange={this.onDisplayTypeChange}
+          // onDisplayTypeChange={this.onDisplayTypeChange}
```
Comment (Ben-CHATA) [inline src/components/ChatMessage/ChatMessage.js:280]: cleanup

File: src/components/ChataTable/ChataTable.js [hunk 16]
```diff
@@ -174,80 +114,181 @@ export default class ChataTable extends React.Component {
...
+    // if (!this.state.isFiltering && prevState.isFiltering) {
+    //   try {
+    //     // this.setFilterTags()
+    //     this.setDateFilterClickListeners()
+    //   } catch (error) {
+    //     console.error(error)
+    //     this.props.onErrorCallback(error)
+    //   }
+    // }
+  }
```
Comment (Ben-CHATA) [inline src/components/ChataTable/ChataTable.js:164]: can probably just get rid of this as well if you're not using it.

File: src/components/ChataTable/ChataTable.js [hunk 17]
```diff
@@ -174,80 +114,181 @@ export default class ChataTable extends React.Component {
...
+      }
+
+      this.setLoading(false)
+    }
   }
 
   setLoading = (loading) => {
     // Don't update state unnecessarily
     if (loading !== this.state.loading) {
+      // setTimeout(() => {
```
Comment (Ben-CHATA) [inline src/components/ChataTable/ChataTable.js:240]: cleanup

File: src/components/ChataTable/ChataTable.js [hunk 18]
```diff
@@ -174,80 +114,181 @@ export default class ChataTable extends React.Component {
...
+      !this.state.pageLoading &&
+      !this.firstRender &&
+      !this.props.isResizing &&
+      !this.props.isAnimating &&
+      !this.props.hidden
+    ) {
+      const height = Math.floor(getComputedStyle(this.tableContainer)?.height)
+      const width = Math.floor(getComputedStyle(this.tableContainer)?.width)
+      // this.tableContainer.style.flexBasis = height
+      // this.tableContainer.style.height = height
```
Comment (Ben-CHATA) [inline src/components/ChataTable/ChataTable.js:258]: cleanup

File: src/components/ChataTable/ChataTable.js [hunk 19]
```diff
@@ -174,80 +114,181 @@ export default class ChataTable extends React.Component {
...
+
+        if (!isNaN(tableWidth) && tableWidth !== this.tableWidth) {
+          this.tableWidth = tableWidth
+        }
+
+        // if (!isNaN(tableHeight)) {
+        //   this.tableHeight = Math.floor(tableHeight) - 50
+        //   this.tableWidth = Math.floor(tableWidth)
+        //   this.props.onSetTableHeight(tableHeight)
+        // }
```
Comment (Ben-CHATA) [inline src/components/ChataTable/ChataTable.js:291]: cleanup

File: src/components/ChataTable/ChataTable.js [hunk 20]
```diff
@@ -437,20 +483,28 @@ export default class ChataTable extends React.Component {
   }
 
   setDateFilterClickListeners = () => {
-    const columns = this.ref?.table?.getColumnDefinitions()
+    // const columns = this.ref?.tabulator?.getColumnDefinitions()
```
Comment (Ben-CHATA) [inline src/components/ChataTable/ChataTable.js:486]: cleanup

File: src/components/ChataTable/ChataTable.js [hunk 21]
```diff
@@ -612,19 +680,34 @@ export default class ChataTable extends React.Component {
   }
 
   getTableHeight = () => {
+    // if (!this.tabulatorMounted || this.props.isAnimating) {
+    //   return '100vh'
+    // } else
```
Comment (Ben-CHATA) [inline src/components/ChataTable/ChataTable.js:685]: cleanup

File: src/components/ChataTable/ChataTable.js [hunk 22]
```diff
@@ -679,8 +762,44 @@ export default class ChataTable extends React.Component {
...
   }
 
+  getFilteredTabulatorColumnDefinitions = () => {
+    if (this.props.columns?.length) {
+      return this.props.columns.map((col) => {
+        const newCol = {}
+        Object.keys(col).forEach((option) => {
+          // if (option === 'headerFilter' && !this.state.isFiltering) {
+          //   newCol.headerFilter = false
+          // } else
```
Comment (Ben-CHATA) [inline src/components/ChataTable/ChataTable.js:772]: cleanup

File: src/components/ChataTable/ChataTable.scss [hunk 23]
```diff
@@ -31,240 +149,184 @@
...
+// .react-autoql-table .tabulator-tableholder:hover::-webkit-scrollbar-track,
+// .react-autoql-table .tabulator-tableholder:hover::-webkit-scrollbar-track-piece,
+// .react-autoql-table .tabulator-tableholder::-webkit-scrollbar-corner {
+//   background-color: transparent;
+// }
+// .react-autoql-table .tabulator-tableholder:hover::-webkit-scrollbar-thumb {
+//   background-color: rgba(0, 0, 0, 0.2);
+//   border-radius: 7px;
+//   border: 0px;
+// }
```
Comment (Ben-CHATA) [inline src/components/ChataTable/ChataTable.scss:332]: cleanup

File: src/components/ChataTable/TableWrapper.js [hunk 24]
```diff
@@ -1,44 +1,146 @@
...
+  tabulator = null //variable to hold your table
+  redrawBlocked = false
+  hasInstantiated = false
+  COMPONENT_KEY = uuid()
 
-    this.isTableMounted = false
+  defaultOptions = {
+    // height: '90%',
+    // maxHeight: '100%',
+    // renderVerticalBuffer: 10,
```
Comment (Ben-CHATA) [inline src/components/ChataTable/TableWrapper.js:21]: cleanup

File: src/components/CustomScrollbars/CustomScrollbars.scss [hunk 25]
```diff
@@ -1,7 +1,7 @@
 ::-webkit-scrollbar {
-  visibility: hidden;
-  width: 6px;
-  height: 6px;
+  // visibility: hidden;
+  // width: 6px;
+  // height: 6px;
```
Comment (Ben-CHATA) [inline src/components/CustomScrollbars/CustomScrollbars.scss:4]: cleanup maybe?

File: src/components/Dashboard/DashboardTile/DashboardTile.js [hunk 26]
```diff
@@ -136,22 +127,32 @@ export class DashboardTile extends React.Component {
       this.setState({ title: _get(this.props, 'tile.title') })
     }
 
-    if (this.secondOptionsToolbarRef?._isMounted) {
-      this.secondOptionsToolbarRef.forceUpdate()
-    }
+    // if (this.secondOptionsToolbarRef?._isMounted) {
+    //   this.secondOptionsToolbarRef.forceUpdate()
+    // }
```
Comment (Ben-CHATA) [inline src/components/Dashboard/DashboardTile/DashboardTile.js:132]: cleanup

File: src/components/Dashboard/DashboardTile/DashboardTile.js [hunk 27]
```diff
@@ -136,22 +127,32 @@ export class DashboardTile extends React.Component {
...
 
-    if (this.optionsToolbarRef?._isMounted) {
-      this.optionsToolbarRef.forceUpdate()
-    }
+    // if (this.optionsToolbarRef?._isMounted) {
+    //   this.optionsToolbarRef.forceUpdate()
+    // }
```
Comment (Ben-CHATA) [inline src/components/Dashboard/DashboardTile/DashboardTile.js:136]: cleanup

File: src/components/Dashboard/DashboardTile/DashboardTile.js [hunk 28]
```diff
@@ -563,7 +566,7 @@ export class DashboardTile extends React.Component {
             className={`dashboard-tile-input-container
             ${this.state.isQueryInputFocused ? 'query-focused' : ''}
             ${this.state.isTitleInputFocused ? 'title-focused' : ''}`}
-            onMouseDown={(e) => e.stopPropagation()}
+            // onMouseDown={(e) => e.stopPropagation()}
```
Comment (Ben-CHATA) [inline src/components/Dashboard/DashboardTile/DashboardTile.js:569]: cleanup

File: src/components/Dashboard/DashboardTile/DashboardTile.js [hunk 29]
```diff
@@ -645,14 +656,14 @@ export class DashboardTile extends React.Component {
             </div>
           </div>
           <div
-            onMouseDown={(e) => e.stopPropagation()}
+            // onMouseDown={(e) => e.stopPropagation()}
```
Comment (Ben-CHATA) [inline src/components/Dashboard/DashboardTile/DashboardTile.js:659]: cleanup

File: src/components/Dashboard/DashboardTile/DashboardTile.js [hunk 30]
```diff
@@ -645,14 +656,14 @@ export class DashboardTile extends React.Component {
...
             className={`dashboard-tile-play-button${!this.isQueryValid(this.state.query) ? ' disabled' : ''}`}
           >
             <Icon type='play' onClick={() => this.processTile()} data-tip='Run tile' data-place='left' />
           </div>
           <div
             className='dashboard-tile-delete-button'
-            onMouseDown={(e) => e.stopPropagation()}
+            // onMouseDown={(e) => e.stopPropagation()}
```
Comment (Ben-CHATA) [inline src/components/Dashboard/DashboardTile/DashboardTile.js:666]: cleanup

File: src/components/Dashboard/DashboardTile/DashboardTile.js [hunk 31]
```diff
@@ -1109,31 +1132,54 @@ export class DashboardTile extends React.Component {
       ${this.props.isEditing ? ' editing' : ''}
       ${this.props.tile.h < 4 ? ' small' : ''}`}
       >
-        <div onMouseDown={(e) => e.stopPropagation()} className='dashboard-tile-response-container'>
+        <div
+          // onMouseDown={(e) => e.stopPropagation()}
```
Comment (Ben-CHATA) [inline src/components/Dashboard/DashboardTile/DashboardTile.js:1136]: cleanup

File: src/components/DataMessenger/DataMessenger.scss [hunk 32]
```diff
@@ -308,7 +307,7 @@
   border: 1px solid #ababab52;
   border-radius: 4px;
   box-shadow: 0px 0 8px var(--react-autoql-box-shadow-color, rgba(0, 0, 0, 0.15));
-  transition: opacity 0.2s ease 0s !important;
+  // transition: opacity 0.2s ease 0s !important;
```
Comment (Ben-CHATA) [inline src/components/DataMessenger/DataMessenger.scss:310]: cleanup

File: src/components/OptionsToolbar/OptionsToolbar.js [hunk 33]
```diff
@@ -624,10 +639,15 @@ export class OptionsToolbar extends React.Component {
   }
 
   render = () => {
-    const shouldShowButton = this.getShouldShowButtonObj()
+    // const displayType = this.props.responseRef?.state?.displayType
+    // if (displayType === 'help' || displayType === 'suggestion') {
+    //   return null
+    // }
```
Comment (Ben-CHATA) [inline src/components/OptionsToolbar/OptionsToolbar.js:645]: cleanup

File: src/components/QueryOutput/QueryOutput.js [hunk 34]
```diff
@@ -484,7 +549,7 @@ export class QueryOutput extends React.Component {
       // Change table ID so a new ChataTable mounts after column change
       // An alternative would be to manually set the new columns in tabulator:
       // this.tableRef.ref.table.setColumns(columns)
-      this.tableID = uuid()
+      // this.tableID = uuid()
```
Comment (Ben-CHATA) [inline src/components/QueryOutput/QueryOutput.js:552]: cleanup

File: src/components/QueryOutput/QueryOutput.js [hunk 35]
```diff
@@ -1713,65 +1849,70 @@ export class QueryOutput extends React.Component {
...
+          originalQueryID={this.props.originalQueryID}
+          isDrilldown={this.isDrilldown()}
+          isQueryOutputMounted={this._isMounted}
+          popoverParentElement={this.props.popoverParentElement}
+          hidden={this.state.displayType !== 'table'}
+          totalRows={this.queryResponse?.data?.data?.count_rows}
+          // onSetTableHeight={(height) => {
+          //   this.tableHeight = height
+          // }}
+          // height={this.tableHeight}
```
Comment (Ben-CHATA) [inline src/components/QueryOutput/QueryOutput.js:1903]: cleanup

File: src/components/Charts/Legend/Legend.js (hunk 8 above)
Comment (NikkiMeganMoore) [reply src/components/Charts/Legend/Legend.js:355]: This one I'll leave in but I left a note

**Review by Ben-CHATA — APPROVED** [review]

**Pull request merged**